### ISSUE TITLE
create multiple new assets at once

### ISSUE THREAD

**Issue by aghisla** [psyinfra/onyo#147]
One of our common use cases (and where we avoid a ton of typing + typos) is batch creation/clone of assets, for example when we get several (from 2 to 10) identical laptops or screens.

In GLPI we solved this by cloning an existing device and editing the clones one by one, replacing information that should be unique (serial number and later inventory number), but it was the best GLPI could offer and it has no validation on unique metadata, which is problematic.

For onyo, Jan and I have initially thought of an `onyo clone` command with the following options:

- asset to be cloned
- how many new clones
- where to put the clones (default: shelf)
- open each new clone in an editor (default) or quick-edit (onyo prompts only for the serial number of each clone)

but we are open to other workflows+implementations that are as linear as this one.

I thought for example about `onyo new` and a almost partially filled template, but I have not yet really decided if templates should only contain keys, at least the fundamental templates - and how onyo should deal with partially filled templates.

**Comment by aqw** [comment]
This is a use case Tobias and I have talked about a bit, but havn't yet come up with a clear answer.

The main topics have been around:

Use cases:
1) import a large database (such as GLPI->onyo). This is likely a one-time event (`onyo import`). This becomes complicated, because it spans across locations and asset types. This, in and of itself isn't a problem, but validation could make the user's experience of `import`ing 1,000 assets painful. How does one skip, resume, etc.
2) add 1 new asset that is very similar to another asset (`onyo clone`). This overlaps quite a bit with templates, and I don't think will be that useful/ergonomic, because a user would have to find another asset somewhere in the tree.
3) add 10 identical (except for serial) laptops (`new --batch`). `onyo new` also allows piping into it (albeit not very ergonomically). So the current solution to this problem is to create a template, create a list of serial numbers, loop over that list and pipe into `onyo new` 10 times. An alternative is to add an option (say `--batch`, `--bulk`, or `--csv`) to `onyo new` to allow multiple assets to be creating in one invocation. This is very, very similar to `onyo import`, but can be interesting with `onyo new --template`.

I currently find option 3 to be most flexible, but the way it will work is still not entirely clear to me. Should `onyo new --batch` create 10 commits, or 1? Should it allow using a template (pre-filling out some fields), and if so, should it allow specifying the template in the CSV file to allow for complicate situations (thus satisfying much of the `onyo import` scenario)?

**Comment by aghisla** [comment]
**TL;DR:** we vote for `onyo new -n N --template TEMPLATE DIR` and use a partially filled template with common values for all new assets. One commit preferred, N commits only if necessary for clean tracking.

We gave some thought to the 3 points and our considerations are:

- 1 (`onyo import`) is not what is meant in this case, and we prefer to interactively type the small amount of unique metadata rather than formatting them in a file for import. I see import as a topic on its own that has, apart from validation, a lot of other aspects to consider, and not a command that we need right now (but would welcome it in the 1.0 release).
- 2 (`onyo clone`) could have its own use cases, but as you said, it may not be so practical. Note that the use of cloning to create one or more new assets is a stretch that we have inherited from GLPI, which means that for Onyo we would both use `clone` for some different action (that we won't discuss here), and associate `new` to our use case.
- 3 (`onyo new -n`) is what would be best suited for our use case, with N as number of new assets to create at once.

The existing `--template` and `DIR` would be applied to all new assets. If we could provide a partially filled template, onyo would ask only for blank fields of that template. That would be transparent for onyo and straightforward for the user.

Regarding one or N commits, I would prefer one (it is one real-life action that deserves one single message), unless it messes up with how git tracks each asset (which I think it will not, but I would make sure).

**Comment by TobiasKadelka** [comment]
This issue will be solved  and closed together with #243, the call will look like:
```onyo new <dir>/<new_asset> <dir>/<new_asset 2> ... <dir>/<new_asset n>```, and alternatively offer an import for many different assets per tsv table.
